### PR TITLE
[flink] add `drop partition`  for flink sql

### PR DESCRIPTION
**Pull request by zhangjun0x01**
### Purpose

add `drop partition`  for flink sql 

### Tests

org.apache.paimon.flink.CatalogTableITCase#testDropPartition

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 1]
```diff
@@ -842,8 +851,23 @@ public final void createPartition(
     @Override
     public final void dropPartition(
             ObjectPath tablePath, CatalogPartitionSpec partitionSpec, boolean ignoreIfNotExists)
-            throws CatalogException {
-        throw new UnsupportedOperationException();
+            throws PartitionNotExistException, CatalogException {
```
Comment (zhuangchong) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:854]: I understand that paimon catalog should be used to drop partition. For example, when using paimon hive catalog, hive metastore needs to be updated at the same time.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 2]
```diff
@@ -842,8 +851,23 @@ public final void createPartition(
...
+            Identifier identifier = toIdentifier(tablePath);
+            Table table = catalog.getTable(identifier);
+            AbstractFileStoreTable fileStoreTable = (AbstractFileStoreTable) table;
+            FileStoreCommit commit = fileStoreTable.store().newCommit(UUID.randomUUID().toString());
+            List<Map<String, String>> partitions =
+                    Collections.singletonList(partitionSpec.getPartitionSpec());
+            commit.dropPartitions(partitions, BatchWriteBuilder.COMMIT_IDENTIFIER);
+        } catch (Catalog.TableNotExistException e) {
+            throw new CatalogException(e);
+        }
```
Comment (zhuangchong) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:870]: I understand that paimon catalog should be used to drop partition. For example, when using paimon hive catalog, hive metastore needs to be updated at the same time.

Comment (zhangjun0x01) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:870]: when we create a paimon partitioned table with hive catalog , [we do not save the partitions info in the hive catalog](https://github.com/apache/incubator-paimon/blob/fa68a1e91ee77aeef0c0b99160cc093e56d2405a/paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java#L455) , so when we drop table partition, we only need to modify the paimon metadata.

Comment (zhuangchong) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:870]: It's handled here.


https://github.com/apache/incubator-paimon/blob/fa68a1e91ee77aeef0c0b99160cc093e56d2405a/paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java#L479-L496

Comment (zhangjun0x01) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java]: yes， you are right, I ignored this parameter `METASTORE_PARTITIONED_TABLE`, I update it

File: paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java [hunk 3]
```diff
@@ -113,6 +118,16 @@ public void createDatabase(String name, boolean ignoreIfExists)
         createDatabaseImpl(name);
     }
 
+    @Override
+    public void dropPartition(Identifier identifier, Map<String, String> partitionSpec)
+            throws TableNotExistException, PartitionNotExistException {
+        Table table = getTable(identifier);
+        AbstractFileStoreTable fileStoreTable = (AbstractFileStoreTable) table;
+        FileStoreCommit commit = fileStoreTable.store().newCommit(UUID.randomUUID().toString());
+        commit.dropPartitions(
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/catalog/AbstractCatalog.java:127]: It looks that we should delete partitions of metastore in `FileStoreCommit`.

**Comment by JingsongLi** [comment]
I think maybe we should not delete partition for metastore, this can be finished by users.
Because we have time travel, the partitions are still visible.

**Comment by zhangjun0x01** [comment]
> I think maybe we should not delete partition for metastore, this can be finished by users. Because we have time travel, the partitions are still visible.

I removed the code follow your suggestion.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 4]
```diff
@@ -826,7 +826,12 @@ public final CatalogPartition getPartition(
     @Override
     public final boolean partitionExists(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)
             throws CatalogException {
-        return false;
+        try {
+            List<CatalogPartitionSpec> partitionSpecs = getPartitionSpecs(tablePath, partitionSpec);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:830]: Here we need an effective method to check partition. For example, we can provide a `SnapshotReaderImpl.withPartitionFilter` to optimize scanning.

Comment (zhangjun0x01) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:823]: I updated it

**Comment by yuzelin** [comment]
Resolve conflict.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**